### PR TITLE
[DeviceMesh] Add get_local_rank() API to DeviceMesh

### PR DESCRIPTION
**Pull request by wz337**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114710
* __->__ #114709
* #114708

As title.

Differential Revision: [D51625152](https://our.internmc.facebook.com/intern/diff/D51625152/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

### CODE REVIEW

File: torch/distributed/_device_mesh.py [hunk 1]
```diff
@@ -381,8 +381,35 @@ def shape(self) -> Tuple[int, ...]:
...
+        """
+        Returns the current global rank.
+        """
         return get_rank()
 
+    def get_local_rank(
+        self, mesh_dim: Optional[Union[int, str]] = None
+    ) -> int:
+        """
+        Returns the local rank of the given mesh_dim.
```
Comment (wanchaol) [inline torch/distributed/_device_mesh.py:393]:
```suggestion
        Returns the local rank on the given mesh_dim of the DeviceMesh. 
```

File: torch/distributed/_device_mesh.py [hunk 2]
```diff
@@ -381,8 +381,35 @@ def shape(self) -> Tuple[int, ...]:
...
+
+        Optional Args:
+            mesh_dim (str/int): it can be the name of the mesh dimension or the index
+            of the mesh dimension. Default is None.
+
+        Returns:
+            An integer denotes the local rank.
```
Comment (wanchaol) [inline torch/distributed/_device_mesh.py:400]: shall we add a small example in the doc about a 2-D mesh, to describe the meaning of local_rank on mesh dimension?

File: torch/distributed/_device_mesh.py [hunk 3]
```diff
@@ -381,8 +381,35 @@ def shape(self) -> Tuple[int, ...]:
...
+        """
+        if self.ndim > 1 and mesh_dim is None:
+            raise RuntimeError(
+                f"Found mesh_dim = {self.mesh.ndim}",
```
Comment (wanchaol) [inline torch/distributed/_device_mesh.py:404]:
```suggestion
                f"Found the DeviceMesh have {self.mesh.ndim} dimensions", 
```

File: torch/distributed/_device_mesh.py [hunk 4]
```diff
@@ -381,8 +381,35 @@ def shape(self) -> Tuple[int, ...]:
...
+                "Optional kwarg `mesh_dim` needs to be specified when mesh_dim > 1."
```
Comment (wanchaol) [inline torch/distributed/_device_mesh.py:405]:
```suggestion
                "Optional kwarg `mesh_dim` needs to be specified when device_mesh.ndim > 1."
```

**Review by wanchaol — APPROVED** [review]
lgtm, some comment suggestions

File: torch/distributed/_device_mesh.py [hunk 5]
```diff
@@ -375,8 +375,50 @@ def shape(self) -> Tuple[int, ...]:
...
+        Example (2 host with 4 GPUs each):
+            ```
+            # Let's initialize device mesh with mesh_shape = (2, 4) to represent
+            # the topology of cross-host(dim 0), and within-host (dim 1).
+            mesh_2d = DeviceMesh(device_type="cuda",
+                            mesh=[
+                                [0, 1, 2, 3],
+                                [4, 5, 6, 7]
+                            ])
+            ```
```
Comment (fegin) [inline torch/distributed/_device_mesh.py:403]: hmm, why "```" exist here?

**Review by fegin — APPROVED** [review]

**Pull request closed without merging**